Generalize initial value retrieval of reduction operators

SoilRos opened the pull request:
The initial value retrieval in the reduction operator is generalized to allow the return from a template argument. The reason behind this is that binary operators return type in the standard are deprecated (C++17) and removed (C++20) (e.g., [`std::plus`](https://en.cppreference.com/w/cpp/utility/functional/plus.html)), meaning that the return type cannot be inferred by purely looking at them.

Note: The idea is to make the next OPM release compatible with the next release of DUNE (see https://lists.dune-project.org/pipermail/dune-devel/2025-October/003173.html), which will require C++20.

SoilRos committed:
Generalize initial value retrieval of reduction operators

The initial value retrieval in the reduction operator is generalized to allow the return from a template argument. The reason behind this is that binary operators in the standard like deprecated (C++17) and removed (C++20) the return type from its traits, meaning that the return type cannot be inferred by purely looking at them.

blattms commented on opm/simulators/linalg/ParallelIstlInformation.hpp:
```diff
@@ -257,21 +248,22 @@ class ParallelISTLInformation
             // we rely on IEE 754 for floating point values and use min()
             // for integral types.
-            if( std::is_integral<Result>::value )
+            if( std::is_integral<T>::value )
```
As we are touching this anyway and are beyond g++4.4, maybe std::numeric_limits<T>::lowest() would work now?

SoilRos replied:
I don't see why not. It's done.

blattms reviewed (approved):
One non-blocking nitpick. Will you address that?

SoilRos committed:
Uses lowest() instead of min/max for initial value

Replaces the custom handling (post g++-4.4) of minimum and maximum values for integral and floating-point types with std::numeric_limits::lowest().
